### PR TITLE
Don't hardcode ink sparkle SPIR-V

### LINKED ISSUE #99783: [Material] [InkSparkle] Change SPIR-V byte code from static const to a file to prepare for Impeller
Binary is currently here: https://github.com/material-components/material-components-flutter-experimental/tree/fragment-shader-manager/fragment_shader_manager/shaders

### PR DESCRIPTION
**Pull request by zanderso**
Fixes #99783

### CODE REVIEW

**Comment by zanderso** [comment]
@dnfield @jonahwilliams About the failing framework_tests_misc and framework_tests_libraries, do you happen to know how to get them to load the shader asset?

**Comment by zanderso** [comment]
Also cc @goderbauer @clocksmith 

This also still needs tool tests.

**Comment by jonahwilliams** [comment]
We'd have to have each test manually load the asset. You could do that by overriding the root bundle and loading the bytes in with synchronous I/O. Unless we have a helper somewhere I've forgotten about

```
Uint8List data;
FakeAssetBundle  bundle;

setUpAll(() {
  data = File('assets/ink_sparkle.spriv).readAsBytesSync();
});

class FakeAssetBundle implements AssetBundle {
  FakeAssetBundle(this.data);
  
  final Uint8List data;
  
  ... override some methods...
}


testWidgets('test', (WidgetTester tester) {
  DefaultAssetBundle(... provide your bundle here);
});
```

**Comment by jonahwilliams** [comment]
The flutter tool can actually build an asset bundle for the test runner, but the problems are that this doesn't go down the same exact path as copyAssets and that technically the flutter framework doesn't have use-material-design on - since that would imply all apps that use flutter would use material design. Anyway, here is a small diff to work from, this outputs the compiled spirv in an easier to get to directory which you might be able to get at with regular asset loading:

```
diff --git a/packages/flutter_tools/lib/src/asset.dart b/packages/flutter_tools/lib/src/asset.dart
index 1bbb098d05..195d6e8f57 100644
--- a/packages/flutter_tools/lib/src/asset.dart
+++ b/packages/flutter_tools/lib/src/asset.dart
@@ -410,8 +410,7 @@ class ManifestAssetBundle implements AssetBundle {
       }
     }
     final List<_Asset> materialAssets = <_Asset>[
-      if (flutterManifest.usesMaterialDesign)
-        ..._getMaterialAssets(),
+      ..._getMaterialAssets(),
     ];
     for (final _Asset asset in materialAssets) {
       final File assetFile = asset.lookupAssetFile(_fileSystem);
diff --git a/packages/flutter_tools/lib/src/build_system/targets/assets.dart b/packages/flutter_tools/lib/src/build_system/targets/assets.dart
index eaf4c7db5c..f6bd39660f 100644
--- a/packages/flutter_tools/lib/src/build_system/targets/assets.dart
+++ b/packages/flutter_tools/lib/src/build_system/targets/assets.dart
@@ -112,7 +112,6 @@ Future<Depfile> copyAssets(
           ) && !await shaderCompiler.compileShader(
             input: content.file as File,
             outputPath: file.path,
-            relativePath: entry.key,
           )) {
             await (content.file as File).copy(file.path);
           }
diff --git a/packages/flutter_tools/lib/src/build_system/targets/shader_compiler.dart b/packages/flutter_tools/lib/src/build_system/targets/shader_compiler.dart
index b09b019f98..e73409339e 100644
--- a/packages/flutter_tools/lib/src/build_system/targets/shader_compiler.dart
+++ b/packages/flutter_tools/lib/src/build_system/targets/shader_compiler.dart
@@ -53,7 +53,6 @@ class ShaderCompiler {
   Future<bool> compileShader({
     required File input,
     required String outputPath,
-    required String relativePath,
   }) async {
        if (!input.path.endsWith('.frag')) {
                return false;
diff --git a/packages/flutter_tools/lib/src/bundle_builder.dart b/packages/flutter_tools/lib/src/bundle_builder.dart
index 2ed5e99387..9a5f2f1072 100644
--- a/packages/flutter_tools/lib/src/bundle_builder.dart
+++ b/packages/flutter_tools/lib/src/bundle_builder.dart
@@ -13,6 +13,7 @@ import 'build_info.dart';
 import 'build_system/build_system.dart';
 import 'build_system/depfile.dart';
 import 'build_system/targets/common.dart';
+import 'build_system/targets/shader_compiler.dart';
 import 'bundle.dart';
 import 'cache.dart';
 import 'devfs.dart';
@@ -148,6 +149,13 @@ Future<void> writeBundle(
   }
   bundleDir.createSync(recursive: true);

+  final ShaderCompiler shaderCompiler = ShaderCompiler(
+    processManager: globals.processManager,
+    logger: globals.logger,
+    fileSystem: globals.fs,
+    artifacts: globals.artifacts!,
+  );
+
   // Limit number of open files to avoid running out of file descriptors.
   final Pool pool = Pool(64);
   await Future.wait<void>(
@@ -161,7 +169,15 @@ Future<void> writeBundle(
         // and the native APIs will look for files this way.
         final File file = globals.fs.file(globals.fs.path.join(bundleDir.path, entry.key));
         file.parent.createSync(recursive: true);
-        await file.writeAsBytes(await entry.value.contentsAsBytes());
+        final DevFSContent devFSContent = entry.value;
+        if (devFSContent is DevFSFileContent) {
+          final File input = devFSContent.file as File;
+          if (!await shaderCompiler.compileShader(input: input, outputPath: file.path)) {
+            input.copySync(file.path);
+          }
+        } else {
+          await file.writeAsBytes(await entry.value.contentsAsBytes());
+        }
       } finally {
         resource.release();
       }
diff --git a/packages/flutter_tools/lib/src/commands/test.dart b/packages/flutter_tools/lib/src/commands/test.dart
index 026529165d..22ad9f197f 100644
--- a/packages/flutter_tools/lib/src/commands/test.dart
+++ b/packages/flutter_tools/lib/src/commands/test.dart
@@ -479,10 +479,10 @@ class TestCommand extends FlutterCommand with DeviceBasedDevelopmentArtifacts {
     if (build != 0) {
       throwToolExit('Error: Failed to build asset bundle');
     }
-    if (_needRebuild(assetBundle.entries)) {
+    // if (_needRebuild(assetBundle.entries)) {
       await writeBundle(globals.fs.directory(globals.fs.path.join('build', 'unit_test_assets')),
           assetBundle.entries);
-    }
+    //}
   }

   bool _needRebuild(Map<String, DevFSContent> entries) {
   ```

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (dnfield) [inline packages/flutter/lib/src/material/ink_sparkle.dart:587]: How confident can we be that this won't get accessed before the future completes?

How important is it to synchronously access this value?

File: packages/flutter/lib/src/material/shaders/ink_sparkle.frag
Comment (dnfield) [inline packages/flutter/lib/src/material/shaders/ink_sparkle.frag:102]: newline at EOF

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (dnfield) [reply packages/flutter/lib/src/material/ink_sparkle.dart:587]: In particular, it lookslike the only actual callsite of this is line 524 above which is in an async method already. We should either expose this as a future or just remove it.

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (dnfield) [inline packages/flutter/lib/src/material/ink_sparkle.dart:522]: This still isn't safe but we're no worse off than we used to be. We should file an issue to deprecate this and remove it. It's not safe for anyone to use.

File: packages/flutter/lib/src/material/shaders/ink_sparkle.frag
Comment (dnfield) [inline packages/flutter/lib/src/material/shaders/ink_sparkle.frag:1]: This should have a copyright header, right?

File: packages/flutter_tools/lib/src/build_system/targets/shader_compiler.dart
Comment (dnfield) [inline packages/flutter_tools/lib/src/build_system/targets/shader_compiler.dart:18]:
```suggestion
  ShaderCompiler({
```

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (zanderso) [reply packages/flutter/lib/src/material/ink_sparkle.dart:522]: @dnfield Can you be more specific about what isn't safe, and what should be removed?

File: packages/flutter_tools/lib/src/build_system/targets/shader_compiler.dart
Comment (dnfield) [inline packages/flutter_tools/lib/src/build_system/targets/shader_compiler.dart:27]:
```suggestion
  }) : _processManager = processManager,
```

File: packages/flutter_tools/lib/src/build_system/targets/shader_compiler.dart
Comment (dnfield) [inline packages/flutter_tools/lib/src/build_system/targets/shader_compiler.dart:72]: File or link bugs

**Review by dnfield — COMMENTED** [review]
test?

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (dnfield) [reply packages/flutter/lib/src/material/ink_sparkle.dart:522]: Sorry, providing sync access to a late initialized variable that gets initialized in a future.

So if some program invokes the getter below here that refers to this static late var before the future completes, it will get a late initialization exception and it has no way to know if that getter is safe to call yet or not.

**Comment by zanderso** [comment]
This PR is blocked on https://github.com/flutter/flutter/issues/102805

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (zanderso) [reply packages/flutter/lib/src/material/ink_sparkle.dart:522]: I think that depends on whether the size of the blobs means that they need to be read from the asset bundle asynchronously. I'm not sure how big e.g. the msl blobs will be but the spir-v for the sparkle shader is already ~11KB.

Comment (jonahwilliams) [reply packages/flutter/lib/src/material/ink_sparkle.dart:522]: I guess technically we should be doing that async anyway to avoid jank on low end devices

Comment (goderbauer) [reply packages/flutter/lib/src/material/ink_sparkle.dart:522]: @dnfield What is the unsafe code pass? The only way to obtain an instance of this is to await `inkSparkle` "static factory" above. At that point, it is guaranteed that _program is initialized, no?

(Side-node: I think `compile` should be a private method since it should only be invoked by `inkSparkle`?)

Comment (goderbauer) [reply packages/flutter/lib/src/material/ink_sparkle.dart:522]: (we could assert that _program is initialized in the inkSparkle method before the return)

Comment (dnfield) [reply packages/flutter/lib/src/material/ink_sparkle.dart:522]: Ahhhh I see, in that case this is fine. I missed that part because of the way GitHub collapsed part of the diff...

**Comment by zanderso** [comment]
Tests are still failing on Windows. Another fix to fml for Windows is here https://github.com/flutter/engine/pull/33029.

**Review by dnfield — APPROVED** [review]
LGTM as long as tests pass

**Comment by zanderso** [comment]
There will be non-trivial work to do this in the internal build. I filed b/231004915 to track that. I am holding off on landing this until the internal roll catches up a bit, and I get feedback on the internal issue from the internal build team.

File: packages/flutter/lib/src/material/shaders/ink_sparkle.frag
Comment (goderbauer) [reply packages/flutter/lib/src/material/shaders/ink_sparkle.frag:1]: Ideally, we also extend the license header check script that warns if a file of this type is added without copyright header.

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (goderbauer) [inline packages/flutter/lib/src/material/ink_sparkle.dart:498]: We should just update the `fragment_shader_manager` to generate the right code. AFAICT, it's only used for this purpose and nothing else?

**Comment by zanderso** [comment]
The g3 side of this is in cl/446113910.

File: packages/flutter/lib/src/material/ink_sparkle.dart
Comment (zanderso) [reply packages/flutter/lib/src/material/ink_sparkle.dart:498]: Yeah, we'll probably want to have a code gen solution, but it should likely be based on the reflection info emitted by impellerc and not on that package. So, spot fixes to `fragment_shader_manager` will probably not actually be too helpful.

**Comment by zanderso** [comment]
Size reductions: https://golem.corp.goog/PerformanceChanges?repository=flutter&revision=130917

**Pull request merged**
